airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java

```diff
@@ -0,0 +1,235 @@
+          final UUID workspaceId,
+          final JsonNode fullConfig,
+          final ConnectorSpecification spec) {
+    return split(uuidSupplier, workspaceId, Jsons.emptyObject(), fullConfig, spec.getConnectionSpecification(), new NoOpSecretPersistence()::read);
+  }
+
+  public static SplitSecretConfig splitUpdate(final Supplier<UUID> uuidSupplier,
+                                              final UUID workspaceId,
+                                              final JsonNode oldPartialConfig,
+                                              final JsonNode newFullConfig,
+                                              final ConnectorSpecification spec,
+                                              final ReadOnlySecretPersistence secretReader) {
+    return split(uuidSupplier, workspaceId, oldPartialConfig, newFullConfig, spec.getConnectionSpecification(), secretReader);
+  }
+
+
+  public static JsonNode combine(final JsonNode partialConfig, final SecretPersistence secretPersistence) {
+    final var config = partialConfig.deepCopy();
+
+    // if the entire config is a secret coordinate object
+    if (config.has(COORDINATE_FIELD)) {
+      final var coordinateNode = config.get(COORDINATE_FIELD);
+      final var coordinate = getCoordinateFromTextNode(coordinateNode);
+      return getOrThrowSecretValueNode(secretPersistence, coordinate);
+    }
+
+    // otherwise iterate through all object fields
+    config.fields().forEachRemaining(field -> {
+      final var fieldName = field.getKey();
+      final var fieldNode = field.getValue();
+
+      if (fieldNode instanceof ArrayNode) {
+        for (int i = 0; i < fieldNode.size(); i++) {
+          ((ArrayNode) fieldNode).set(i, combine(fieldNode.get(i), secretPersistence));
+        }
+      } else if (fieldNode instanceof ObjectNode) {
+        ((ObjectNode) config).replace(fieldName, combine(fieldNode, secretPersistence));
+      }
+    });
+
+    return config;
+  }
+
+  private static SplitSecretConfig split(
+          final Supplier<UUID> uuidSupplier,
+          final UUID workspaceId,
+          final JsonNode oldPartialConfig,
+          final JsonNode originalFullConfig,
+          final JsonNode spec,
+          final ReadOnlySecretPersistence secretReader) {
+    Preconditions.checkArgument(JsonSecretsProcessor.canBeProcessed(spec), "Schema is not valid JSONSchema!");
+
+    final var fullConfig = originalFullConfig.deepCopy();
+    final var secretMap = new HashMap<SecretCoordinate, String>();
+    final var specProperties = (ObjectNode) spec.get(JsonSecretsProcessor.PROPERTIES_FIELD);
+
+    for (final String specProperty : Jsons.keys(specProperties)) {
+      final var fieldSchema = specProperties.get(specProperty);
+      final var fieldSchemaCombinationType = JsonSecretsProcessor.findJsonCombinationNode(fieldSchema);
+
+      // if the json schema field is an obj and has the airbyte secret field
+      if (fullConfig.has(specProperty)) {
```

Comment:
I'm a little wary of moving these into private functions because that would hide where the `fullConfig` object is getting updated. I tried restructuring a lot in the other function and outlining what is happening in javadocs.